daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala

```diff
@@ -443,13 +443,13 @@ private[validation] object Typing {
       env.checkExpr(observers, TParties)
       env.checkExpr(agreementText, TText)
       choices.values.foreach(env.checkChoice(tplName, _))
+      env.checkIfaceImplementations(tplName, implementations)
       mbKey.foreach { key =>
         checkType(key.typ, KStar)
         env.checkExpr(key.body, key.typ)
         checkExpr(key.maintainers, TFun(key.typ, TParties))
         ()
       }
-      env.checkIfaceImplementations(tplName, implementations)
```

Comment:
This was buggy -- the implementation had access to a `this` param when it shouldn't have.
Now it should have access to it. 

(I changed the order here so that it matches the Haskell side.)